akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala

```diff
@@ -306,6 +306,67 @@ object GraphStages {
     override def toString: String = s"SingleSource($elem)"
   }
 
+  final class FutureFlattenSource[T, M](
+    val future: Future[Graph[SourceShape[T], M]])
+    extends GraphStageWithMaterializedValue[SourceShape[T], Future[M]] {
+
+    ReactiveStreamsCompliance.requireNonNullElement(future)
+
+    val out = Outlet[T]("futureFlatten.out")
+    val shape = SourceShape(out)
+
+    override def initialAttributes = DefaultAttributes.futureSource // TODO
+
+    def createLogicAndMaterializedValue(attr: Attributes): (GraphStageLogic, Future[M]) = {
+      val materialized = Promise[M]()
+
+      val logic = new GraphStageLogic(shape) with InHandler with OutHandler {
+        private val sinkIn = new SubSinkInlet[T]("FlattenMergeSink")
+
+        private val cb = getAsyncCallback[Try[Graph[SourceShape[T], M]]] {
+          case scala.util.Success(graph) ⇒ {
+            setHandler(out, this)
+            sinkIn.setHandler(this)
+
+            sinkIn.pull()
+
+            val src = Source.fromGraph(graph).watchTermination() { (m, done) ⇒
+              materialized.completeWith(
+                done.map(_ ⇒ m)(ExecutionContexts.sameThreadExecutionContext))
+            }
```

Comment:
This is not correct. The materialized value promise should complete as soon as the inner graph is materialized, not when it completes. 

Extracting the value to complete the promise can be done by just `Keep`ing it when materializing below, and have run return it so that it can be sent to the promise. The promise should be failed if materialization fails.